doc/sfp-cmis/cmis-init.md

```diff
@@ -0,0 +1,301 @@
+            "application_selection": false
+        },
+        "DELL EMC#6MGDY": {
+            "application_selection": true
+        },
+        "DELL EMC#DH11M": {
+            "application_selection": false
+        }
+    }
+}
+```
+  As of now, only **application_selection** is supported, and it defaults to **True** if unspecified
+
+- Add a CMIS manager class to handle the application initializations in parallel  
+
+  ![](images/001.png)
+
+- Add **PortChangeEvent.PORT_SET** and **PortChangeEvent.PORT_DEL** events to **xcvrd_utilities/port_mapping.py**  
+  - PortChangeEvent.PORT_SET  
+    Invoke the event callbacks upon **SET** operations, it includes all the port config updates,
+not just the port add/remove events.
+  - PortChangeEvent.PORT_DEL  
+    Invoke the event callbacks upon **DEL** operations.
+
+- As of now, the number of CmisManagerTask is hard coded to 4  
+  - Task 0 is dedicated to listen for **PortChangeEvent.PORT_SET** events, and put requests into the shared message queue
+  - Task 1-N are responsible for handling the CMIS application initializations in parallel,
+  since all the requests are from the shared message queue, only one request will be processed
+  by a certain task at a time.
+
+  ![](images/002.png)
+
+## sonic-platform-common/sonic_platform_base/sfp_base.py (modified)
+
+Add the following stub routines for the CMIS transceiver
+- get_cmis_application_selected(self)  
+Retrieves the selected application code of this CMIS transceiver
+- get_cmis_application_matched(self, host_speed, host_lanes)  
+Retrieves the matched application code by the host interface config
+- set_cmis_application(self, host_speed, host_lanes)  
+Updates the application selection of this CMIS transceiver
+
+## sonic-platform-common/sonic_platform_base/sonic_xcvr/api/public/cmis.py (modified)
+
+- Add support for software reset
+- Add support for activating the CMIS application base on the host port mode.
+- Detailed CMIS application initialization sequence is available in **Appendix D** of
+[Common Management Interface Specification Rev. 5](http://www.qsfp-dd.com/wp-content/uploads/2021/05/CMIS5p0.pdf)
+
+## sonic-platform-common/sonic_platform_base/sonic_xcvr/fields/consts.py (modified)
+
+- Add register definitions for CMIS application initialization
+
+## sonic-platform-common/sonic_platform_base/sonic_xcvr/fields/public/cmis.py (modified)
+
+- Add support for parsing CMIS application advertisement
+
+## sonic-platform-common/sonic_platform_base/sonic_xcvr/mem_maps/public/cmis.py (modified)
+
+- Add the meory map for the CMIS registers which is necessary for application initialization
+
+## sonic-platform-common/sonic_platform_base/sonic_xcvr/sfp_optoe_base.py (modified)
+
+- Add support for CMIS application initialization
+  - get_cmis_application_selected()
+  - get_cmis_application_matched()
+  - set_cmis_application()
+- Add mutex lock to the follow routines to serialize CMIS application initializations  
```

Comment:
Yes, it will, but the main purpose of this PR is for the dynamic-port-breakout support, which is having all the lanes configured in the same application mode, hence a reset() is a simple and safest way to ensure the entire module is reset to a known state. I'll later update the code to remove the reset()